### PR TITLE
Adding an implementation of multiply for Bit Array

### PR DESCRIPTION
**Pull request by benjaminsavage**
See the super long comment =).

Eventually, this will be used as an "Extension Field" to be able to operate on arithmetic circuits in F2, with any security parameter we choose, as outlined in https://eprint.iacr.org/2019/386.pdf

### CODE REVIEW

File: src/bits/fp2_array.rs [hunk 1]
```diff
@@ -123,10 +125,111 @@ macro_rules! bit_array_impl {
...
+            // 5. Inverse elements: For any non-zero element a in the field, there exists a unique element -a such that a+(-a) = 0, and there exists a unique element a^(-1) such that a×a^(-1) = 1.
+            // 6. Distributivity: Multiplication distributes over addition, meaning that a(b+c) = ab+ac for any elements a, b, and c in the field.
+            //
+            // Tests of Associativity, Commutativity and Distributivity are below.
             impl std::ops::Mul for $name {
                 type Output = Self;
                 fn mul(self, rhs: Self) -> Self::Output {
-                    self & rhs
+                    debug_assert!(2 * $bits < u128::BITS);
+                    let a = <Self as Into<u128>>::into(self);
```
Comment (taikiy) [inline src/bits/fp2_array.rs:219]: Let's change this to
```suggestion
                    let a = <Self as Fp2Array>::as_u128(self);
```

File: src/bits/fp2_array.rs [hunk 2]
```diff
@@ -123,10 +125,111 @@ macro_rules! bit_array_impl {
...
+                    let mut product = 0;
+                    for i in 0..Self::BITS {
+                        let bit = u128::from(rhs[i]);
+                        product ^= bit * (a << i);
+                    }
+
+                    while (u128::BITS - product.leading_zeros()) > $bits {
+                        let poly = <Self as Fp2Array>::POLYNOMIAL;
```
Comment (taikiy) [inline src/bits/fp2_array.rs:227]: Move this to the top of this function?

File: src/bits/fp2_array.rs [hunk 3]
```diff
@@ -335,6 +470,32 @@ macro_rules! bit_array_impl {
...
-bit_array_impl!(bit_array_40, BitArray40, U8_5, 40, U5);
-bit_array_impl!(bit_array_8, BitArray8, U8_1, 8, U1);
+bit_array_impl!(
+    bit_array_40,
+    BitArray40,
+    U8_5,
+    40,
+    U5,
+    // x^40 + x^5 + x^3 + x^2 + 1
+    0b1_0000_0000_0000_0000_0000_0000_0000_0000_0010_1101_u128
```
Comment (taikiy) [inline src/bits/fp2_array.rs:480]: like the underscores 👍

**Review by taikiy — APPROVED** [review]
Just one nit suggestion. LGTM

File: src/bits/fp2_array.rs (hunk 1 above)
Comment (benjaminsavage) [reply src/bits/fp2_array.rs:219]: Why would you suggest this? Why not the more generic Into?

File: src/bits/fp2_array.rs (hunk 2 above)
Comment (benjaminsavage) [reply src/bits/fp2_array.rs:227]: Fair enough!

File: src/bits/fp2_array.rs (hunk 3 above)
Comment (benjaminsavage) [reply src/bits/fp2_array.rs:480]: clippy made me do it =P

**Pull request merged**